code_generation/configuration.py

```diff
@@ -0,0 +1,526 @@
+            parameters: The parameters to be added. This must be a dictionary of parameters. If multiple scopes are given,
+            the parameters are added to all scopes.
+
+        Returns:
+            None
+        """
+        if not isinstance(scopes, list):
+            scopes = [scopes]
+        for scope in scopes:
+            self.config_parameters[scope].update(self.resolve_modifiers(parameters))
+
+    def add_producers(self, scopes: ScopeList, producers: TProducerInput) -> None:
+        """
+        Function used to add producers to the configuration. Internally, a set of all
+        available outputs is updated, which is later used to check if all required ouputs are available.
+
+        Args:
+            scopes: The scopes to which the producers should be added. This can be a list of scopes or a single scope.
+            producers: The producers to be added. This must be a list of producers. If multiple scopes are given,
+            the producers are added to all scopes.
+
+        Returns:
+            None
+        """
+        if not isinstance(scopes, list):
+            scopes = [scopes]
+        if not isinstance(producers, list):
+            producers = [producers]
+        for scope in scopes:
+            self.producers[scope].extend(producers)
+            self.available_outputs[scope].update(
+                CollectProducersOutput(producers, scope)
+            )
+
+    def add_outputs(self, scopes: ScopeList, output: QuantitiesInput) -> None:
+        """
+        Function used to add outputs to the configuration.
+
+        Args:
+            scopes: The scopes to which the outputs should be added. This can be a list of scopes or a single scope.
+            outputs: The outputs to be added. This must be a list of outputs. If multiple scopes are given,
+            the outputs are added to all scopes.
+
+        Returns:
+            None
+        """
+        if not isinstance(scopes, list):
+            scopes = [scopes]
+        if not isinstance(output, list):
+            output = [output]
+        for scope in scopes:
+            self.outputs[scope].update(output)
+
+    def add_shift(
+        self,
+        shift: Union[SystematicShift, SystematicShiftByQuantity],
+        samples: Union[str, List[str], None] = None,
+    ) -> None:
+        """
+        Function used to add a systematics shift to the configuration. During this step, the shift is validated and applied.
+
+        Args:
+            shift: The shift to be added. This must be a SystematicShift object.
+            samples: The samples to which the shift should be applied. This can be a list of samples or a single sample.
+                If ths option is not set, the shift is applied, regardless of the sample type.
+
+        Returns:
+            None
+        """
+        if not isinstance(shift, SystematicShift):
```

Comment:
This means `SystematicShiftByQuantity` can't  be used yet?

Reply:
Same as for the Quantitiy, the `SystematicShiftByQuantity` class inherits from `SystematicShift`so the isInstance call is also true for `SystematicShiftByQuantity`

https://github.com/KIT-CMS/CROWN/blob/eff8756bcf070769078e2c7ed6d840eaac37f1e1/code_generation/systematics.py#L401